prefect_docker/worker.py

```diff
@@ -0,0 +1,607 @@
+            )
+        finally:
+            docker_client.close()
+
+        return container
+
+    def _watch_container(
+        self,
+        docker_client: "DockerClient",
+        container_id: str,
+        configuration: DockerWorkerJobConfiguration,
+    ) -> Generator[None, None, "Container"]:
+        """
+        Watches a container for completion, yielding the latest container
+        snapshot on each iteration.
+        """
+        container: "Container" = docker_client.containers.get(container_id)
+
+        status = container.status
+        self._logger.info(
+            f"Docker container {container.name!r} has status {container.status!r}"
+        )
+        yield container
+
+        if configuration.stream_output:
+            try:
+                for log in container.logs(stream=True):
+                    log: bytes
+                    print(log.decode().rstrip())
+            except docker.errors.APIError as exc:
+                if "marked for removal" in str(exc):
+                    self._logger.warning(
+                        f"Docker container {container.name} was marked for removal"
+                        " before logs could be retrieved. Output will not be"
+                        " streamed. "
+                    )
+                else:
+                    self._logger.exception(
+                        "An unexpected Docker API error occured while streaming output "
+                        f"from container {container.name}."
+                    )
+
+            container.reload()
+            if container.status != status:
+                self._logger.info(
+                    f"Docker container {container.name!r} has status"
+                    f" {container.status!r}"
+                )
+            yield container
+
+        container.wait()
+        self._logger.info(
+            f"Docker container {container.name!r} has status {container.status!r}"
+        )
+        yield container
+
+    def _should_pull_image(
+        self, docker_client: "DockerClient", configuration: DockerWorkerJobConfiguration
+    ) -> bool:
+        """
+        Decide whether we need to pull the Docker image.
+        """
+        image_pull_policy = configuration._determine_image_pull_policy()
+
+        if image_pull_policy is ImagePullPolicy.ALWAYS:
+            return True
+        elif image_pull_policy is ImagePullPolicy.NEVER:
+            return False
+        elif image_pull_policy is ImagePullPolicy.IF_NOT_PRESENT:
+            try:
+                # NOTE: images.get() wants the tag included with the image
+                # name, while images.pull() wants them split.
+                docker_client.images.get(configuration.image)
+            except docker.errors.ImageNotFound:
+                self._logger.debug(
+                    f"Could not find Docker image locally: {configuration.image}"
+                )
+                return True
+        return False
+
+    def _pull_image(
```

Comment:
I don't think you need to address this in this PR (I know much of this is copy / paste from the old block), but I find classes / files with this many methods to be a mine field; IMO separating something out into its own thing should be done only when that thing is easily testable / reusable as an individual unit, and I think many of these methods (including this one) don't pass that standard

Reply:
I agree that there's room for improvement in the implementation of this class. There are some existing utilities that could be used to replace portions of this class. I've created #37 for tracking.